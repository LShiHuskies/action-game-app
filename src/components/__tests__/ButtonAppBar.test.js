import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';


import {
    Router,
    Switch,
    Route
  } from "react-router-dom";



import ButtonAppBar from '../buttonAppBar';



const history = createMemoryHistory();
const renderComponent = () => {
    const helpers = render(<Router history={history}>
        <Switch>
          <Route> <ButtonAppBar /> </Route>
        </Switch>
      </Router>);

    return { ...helpers };
}



describe('<ButtonAppBar />', () => {
    it('should match snapshot', () => {
        const tree = renderer.create(<Router history={history}>
            <Switch>
              <Route> <ButtonAppBar /> </Route>
            </Switch>
          </Router>);

          expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should have a button on page', () => {
        renderComponent();

        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    describe('when clicking the button', () => {
        it('should trigger the opening with dropdown menu with 3 menuitems', () => {
            renderComponent();
            fireEvent.click(screen.getByRole('button'));

            expect(screen.getByRole('menu')).toBeInTheDocument();
            expect(screen.getByRole('menuitem', { name: 'Profile' })).toBeInTheDocument();
            expect(screen.getByRole('menuitem', { name: 'My account' })).toBeInTheDocument();
            expect(screen.getByRole('menuitem', { name: 'Logout' })).toBeInTheDocument();
        });


        describe('after clicking profile selection', () => {
        
            describe('when the current pathnaame is already profile', () => {
                let location;
                const mockLocation = new URL("https://example.com/profile");
                beforeEach(() => {
                    location = window.location;
                    mockLocation.replace = jest.fn();
            
                    delete window.location;
                    window.location = mockLocation;
                });
            
                afterEach(() => {
                    window.location = location;
                });
    
                it('should not push into the history', () => {
                    renderComponent();
                    fireEvent.click(screen.getByRole('button'));
                    const previousHistoryLength = history.length;
                    fireEvent.click(screen.getByRole('menuitem', { name: 'Profile' }));
    
                    expect(previousHistoryLength).toEqual(history.length);
                });
            });



            describe('when the current pathname is not already profile', () => {
                let location;
                const mockLocation = new URL("https://example.com/something");
                let originalError;
                beforeEach(() => {
                    location = window.location;
                    mockLocation.replace = jest.fn();
            
                    delete window.location;
                    window.location = mockLocation;
                    originalError = console.error;
                    console.error = jest.fn();
                });
            
                afterEach(() => {
                    window.location = location;
                    console.error = originalError;
                });

                it('should direct to the profile path', () => {
                    renderComponent();
                    fireEvent.click(screen.getByRole('button'));
    
                    fireEvent.click(screen.getByRole('menuitem', { name: 'Profile' }));

                    expect(history.length).toBe(2);
                });
            });


        });

    });


});
