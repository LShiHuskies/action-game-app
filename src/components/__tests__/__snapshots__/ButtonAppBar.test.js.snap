// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonAppBar /> should match snapshot 1`] = `
Array [
  " ",
  <div
    className="css-e53awj-MuiStack-root"
    style={
      Object {
        "backgroundColor": "rgba(242, 121, 53, 1)",
        "height": "50px",
        "justifyContent": "flex-end",
      }
    }
  >
    <button
      aria-haspopup="true"
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="composition-button"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        style={
          Object {
            "color": "#282c34",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
  </div>,
  " ",
]
`;
